x/lockup/keeper/msg_server_test.go

```diff
@@ -9,55 +9,77 @@ import (
 	"github.com/osmosis-labs/osmosis/x/lockup/types"
 )
 
-// TODO: Make table driven
 func (suite *KeeperTestSuite) TestMsgLockTokens() {
-	suite.SetupTest()
-
-	// lock coins
-	addr1 := sdk.AccAddress([]byte("addr1---------------"))
-	coins := sdk.Coins{sdk.NewInt64Coin("stake", 10)}
-
-	err := simapp.FundAccount(suite.app.BankKeeper, suite.ctx, addr1, coins)
-	suite.Require().NoError(err)
-	_, err = suite.app.LockupKeeper.LockTokens(suite.ctx, addr1, coins, time.Second)
-	suite.Require().NoError(err)
-
-	// creation of lock via LockTokens
-	msgServer := keeper.NewMsgServerImpl(*suite.app.LockupKeeper)
-	_, err = msgServer.LockTokens(sdk.WrapSDKContext(suite.ctx), types.NewMsgLockTokens(addr1, time.Second, coins))
-
-	// check locks
-	locks, err := suite.app.LockupKeeper.GetPeriodLocks(suite.ctx)
-	suite.Require().NoError(err)
-	suite.Require().Len(locks, 1)
-	suite.Require().Equal(locks[0].Coins, coins)
-	// check accumulation store is correctly updated
-	accum := suite.app.LockupKeeper.GetPeriodLocksAccumulation(suite.ctx, types.QueryCondition{
-		LockQueryType: types.ByDuration,
-		Denom:         "stake",
-		Duration:      time.Second,
-	})
-	suite.Require().Equal(accum.String(), "10")
-
-	// add more tokens to lock via LockTokens
-	addCoins := sdk.Coins{sdk.NewInt64Coin("stake", 10)}
-	err = simapp.FundAccount(suite.app.BankKeeper, suite.ctx, addr1, addCoins)
-	suite.Require().NoError(err)
-
-	_, err = msgServer.LockTokens(sdk.WrapSDKContext(suite.ctx), types.NewMsgLockTokens(addr1, locks[0].Duration, addCoins))
-	suite.Require().NoError(err)
-
-	// check locks after adding tokens to lock
-	locks, err = suite.app.LockupKeeper.GetPeriodLocks(suite.ctx)
-	suite.Require().NoError(err)
-	suite.Require().Len(locks, 1)
-	suite.Require().Equal(locks[0].Coins, coins.Add(addCoins...))
-
-	// check accumulation store is correctly updated
-	accum = suite.app.LockupKeeper.GetPeriodLocksAccumulation(suite.ctx, types.QueryCondition{
-		LockQueryType: types.ByDuration,
-		Denom:         "stake",
-		Duration:      time.Second,
-	})
-	suite.Require().Equal(accum.String(), "20")
+	type param struct {
+		coinsToLock sdk.Coins
+		lockOwner   sdk.AccAddress
+		duration    time.Duration
+	}
+
+	tests := []struct {
+		name       string
+		param      param
+		expectPass bool
+	}{
+		{
+			name: "creation of lock via lockTokens",
+			param: param{
+				coinsToLock: sdk.Coins{sdk.NewInt64Coin("stake", 10)},       // setup wallet
+				lockOwner:   sdk.AccAddress([]byte("addr1---------------")), // setup wallet
+				duration:    time.Second,
+			},
+			expectPass: true,
+		},
+	}
+
+	for _, test := range tests {
+		suite.SetupTest()
+
+		if test.expectPass {
+			err := simapp.FundAccount(suite.app.BankKeeper, suite.ctx, test.param.lockOwner, test.param.coinsToLock)
+			suite.Require().NoError(err)
+
+			_, err = suite.app.LockupKeeper.LockTokens(suite.ctx, test.param.lockOwner, test.param.coinsToLock, test.param.duration)
```

Comment:
@ValarDragon @mattverse 
Why doesn't `suite.app.LockupKeeper.LockTokens(...)` throw an error when my coinsToLock > coinsInAddress? 

I was expecting an error in this line, but seems like I need to go all the way to line 53 `suite.Require().Equal(locks[0].Coins, test.param.coinsToLock)` to return the error 

Is there a better way to handle this?

Reply:
With what parameters are you trying to create a test? Are you trying to create a test case for trying to lock tokens with more than the account has?

Reply:
Yup

```	{
			name: "locking more coins than are in the address",
			param: param{
				coinsToLock:         sdk.Coins{sdk.NewInt64Coin("stake", 20)},       // setup wallet
				lockOwner:           sdk.AccAddress([]byte("addr1---------------")), // setup wallet
				duration:            time.Second,
				coinsInOwnerAddress: sdk.Coins{sdk.NewInt64Coin("stake", 10)},
			},
			expectPass: false,
		},
```

Reply:
I see, I think it's probably because you're funding the account at line 39 with `test.params.CoinsToLock`, not with `test.params.coinsInOwnerAddress`

Reply:
aah i see

Reply:
Also @mattverse the minimum duration to LockTokens is a day right and anything <=0 should throw an error? 

Can you please direct me to the file where errors are handled for `lockTokens`? I can't seem to figure out if it's done locally or not

Reply:
Actually there isn't a minimum duration for LockTokens, they are completely different from incentvies( as of which has the minimum duration of a day to be eligible for incentives) 

Per different test cases for lockTokens, you might want to take a look at https://github.com/osmosis-labs/osmosis/blob/main/x/lockup/keeper/lock_test.go

Reply:
I see that makes sense. In that case, this PR is ready for full review!